Sources/Crypto/KDF/PBKDF2.swift

```diff
@@ -0,0 +1,181 @@
+import Foundation
+import Core
+
+/// PBKDF2 derives a fixed or custom length key from a password and salt.
+///
+/// It accepts a customizable amount of iterations to increase the algorithm weight and security.
+///
+/// Unlike BCrypt, the salt does not get stored in the final result,
+/// meaning it needs to be generated and stored manually.
+///
+///     let passwordHasher = PBKDF2(digest: SHA1)
+///     let salt = try CryptoRandom().generateData(count: 64) // Data
+///     let hash = try passwordHasher.deriveKey(fromPassword: "secret", salt: salt, iterations: 15_000) // Data
+///     print(hash.hexEncodedString()) // 8e55fa3015da583bb51b706371aa418afc8a0a44
+///
+/// PBKDF2 leans on HMAC for each iteration and can use all hash functions supported in Crypto
+///
+/// https://en.wikipedia.org/wiki/PBKDF2
+public final class PBKDF2 {
+    /// The requested amount of output bytes from the key derivation
+    ///
+    /// In circumstances with low iterations the amount of output bytes may not be met.
+    ///
+    /// `digest.digestSize * iterations` is the amount of bytes stored in PBKDF2's buffer.
+    /// Any data added beyond this limit
+    public enum KeySize {
+        case digestSize
+        case fixed(Int)
+        
+        fileprivate func size(for digest: Digest) -> Int {
+            switch self {
+            case .digestSize:
+                return numericCast(digest.algorithm.digestSize)
+            case .fixed(let size):
+                return size
+            }
+        }
+    }
+    
+    private let digest: Digest
+    
+    /// Creates a new PBKDF2 derivator based on a hashing algorithm
+    public init(digest: Digest) {
```

Comment:
Some conveniences like [this](https://github.com/vapor/crypto/blob/master/Sources/Crypto/MAC/HMAC.swift#L113-L148) would be nice so that people can just do `PBKDF2.SHA256.deriveKey(...)`.